uart/gps_uart/gps_uart.c

```diff
@@ -0,0 +1,251 @@
+
+    bool is_valid = strncmp(hex_checksum, checksum_ptr + 1, 2) == 0;
+    printf("Checksum Validation: %s\n", is_valid ? "VALID" : "INVALID");
+
+    return is_valid;
+}
+
+
+/**
+ * @brief Converts an NMEA coordinate to decimal degrees.
+ *
+ * This function converts an NMEA-formatted coordinate (DDMM.MMMM) to
+ * decimal degrees (DD.DDDDDD).
+ *
+ * @param nmea_coord The NMEA coordinate string to convert.
+ * @param decimal_coord Pointer to store the converted decimal coordinate.
+ */
+void convert_nmea_to_decimal(char *nmea_coord, float *decimal_coord)
+{
+    float degrees = atof(nmea_coord) / 100.0;
+    int int_degrees = (int)degrees;
+    float minutes = (degrees - int_degrees) * 100.0;
+
+    *decimal_coord = int_degrees + (minutes / 60.0);
+}
+
+
+/**
+ * @brief Parses an NMEA sentence and extracts GPS data.
+ *
+ * This function parses an NMEA sentence (currently only $GPRMC) and extracts
+ * latitude, longitude, and validity information. It also validates the checksum.
+ *
+ * @param nmea_string The NMEA sentence to parse.
+ * @param gps_data Pointer to the GPSData structure to store the parsed data.
+ * @return true if the sentence was successfully parsed, false otherwise.
+ */
+bool parse_nmea_gps(char *nmea_string, GPSData *gps_data)
+{
+    if (!validate_nmea_checksum(nmea_string))
+    {
+        printf("Invalid NMEA Checksum\n");
+        return false;
+    }
+
+    // Ignore all NMEA strings that are not of type GPRMC
+    if (strncmp(nmea_string, "$GPRMC", 6) != 0)
+    {
+        return false;
+    }
+    
+    char *token;
+    char *tokens[12] = {0};
+    int token_count = 0;
+
+    // Tokenize the string
+    token = strtok(nmea_string, ",");
+    while (token != NULL && token_count < 12)
+    {
+        tokens[token_count++] = token;
+
+        token = strtok(NULL, ",");
+    }
+
+    // Check if we have enough tokens and data is valid
+    if (token_count >= 12 && strcmp(tokens[2], "A") == 0)
+    {
+        if (strlen(tokens[3]) > 0)
+        {
+            convert_nmea_to_decimal(tokens[3], &gps_data->latitude);
+            if (tokens[4][0] == 'S')
+                gps_data->latitude = -gps_data->latitude;
+        }
+
+        if (strlen(tokens[5]) > 0)
+        {
+            convert_nmea_to_decimal(tokens[5], &gps_data->longitude);
+            if (tokens[6][0] == 'W')
+                gps_data->longitude = -gps_data->longitude;
+        }
+
+        gps_data->is_valid = true;
+        return true;
+    }
+}
+
+
+/**
+ * @brief Processes incoming GPS data from the UART interface.
+ *
+ * This function reads NMEA sentences from the UART interface, validates them,
+ * and updates the GPSData structure if a valid sentence is received.
+ *
+ * @param gps_data Pointer to the GPSData structure to store the parsed data.
+ */
+void process_gps_data(GPSData *gps_data)
+{
+    char nmea_buffer[MAX_NMEA_LENGTH];
+    int chars_read = 0;
+
+    while (uart_is_readable(UART_ID) && chars_read < MAX_NMEA_LENGTH - 1)
```

Comment:
This function doesn't handle uart_is_readable being false before the end of the line